src/ir_passes/eq.ml

```diff
@@ -134,16 +146,22 @@ let eq_for : T.typ -> Ir.dec * T.typ list = fun t ->
       define_eq t (array_eq_func_body t' (varE (eq_var_for t')) (arg1E t) (arg2E t)),
       [t']
     end
-  | T.Obj ((T.Object | T.Memory | T.Module), fs) ->
+  | T.Obj (T.(Object | Memory | Module), fs) ->
     define_eq t (
       conjE (List.map (fun f ->
         let t' = T.as_immut (T.normalize f.Type.typ) in
         eq_func_body t' (dotE (arg1E t) f.Type.lab t') (dotE (arg2E t) f.Type.lab t')
       ) fs)
     ),
     List.map (fun f -> T.as_immut (T.normalize (f.Type.typ))) fs
+  | T.Variant fs when List.for_all is_flat_alt fs ->
+    (* enum-like, i.e. flat variant *)
+    define_eq_variant t (trueE ()),
+    List.map (fun (f : T.field) -> T.normalize f.T.typ) fs
   | T.Variant fs ->
-    define_eq t (
+    let flat, deep = List.partition is_flat_alt fs in
+    assert (deep <> []);
+    define_eq_variant t (
       (* switching on the diagonal *)
       { it = SwitchE
         ( tupE [arg1E t; arg2E t],
```

Comment:
To avoid the tuple, would it be better to generate nested switches?

Reply:
Nope, I think we could use multi-value (`SR.UnboxedTuple 2`) to pass around pairs on stack and keep them in variables.